better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+  public var length: Int { ... }
+    .
+    .
+    .
+
+  /// The number of elements that can be stored before storage is
+  /// reallocated.
+  private var capacity: Int { ... }
+}
+```
+
+The method bodies are collapsed, because contracts are not about what
+you put in your implementations.  Contracts are part of your
+function's interface.
+
+#### Basics
+
+We've started by filling in the most basic part of every documentation
+comment: a summary sentence fragment that minimally describes what the
+thing *is* or *does*.
+
+The first one
+
+```swift
+/// A resizable random-access collection of `T`s.
+struct DynamicArray<T>
+```
+
+gives us the context we need to understand the methods: we're looking
+at the declaration of dynamic array type that holds any number of Ts.
+
+Now let's look at the documentation for the first method, called
+"popLast."
+
+```swift
+  /// Removes and returns the last element.
+  public mutating func popLast() -> T { ... }
+```
+
+As you can see from the summary, it removes and returns the last
+element. Notice that the phrase “last element” is meaningful only
+because we documented that `DynamicArray` is a collection, which we
+know is a sequence of elements.
+
+The summary for a method should tell us what the method returns, and
+what its other effects are, if any.  This method is a little unusual
+in that it both mutates the array and returns a result, which means
+you need to decide what to emphasize in the summary: the removal or
+the returned value.  Here we've emphasized the mutation, which is
+normally what you want. That emphasis is reflected by the name, which
+describes the mutation rather than the return value.
+
+You'd generally only emphasize the return value if the mutation
+is something incidental that doesn't affect the program's meaning,
+like updating a cache.
+
+Notice also that we've omitted needless words: we didn't write, “the
+last element of the `DynamicArray`,” because the type sets the context
+for its methods.  Remember, every word the reader has to process takes
+mental energy and contributes to perceived complexity.  We use all the
+necessary words, but no more than necessary.
+
+#### Details
+
+Let's continue by spelling out the preconditions, postconditions,
+and invariants of `popLast()`.
+
+What are the preconditions for removing an element?  Obviously, there
+needs to be an element to remove.
+
+```swift
+  /// Removes and returns the last element.
+  ///
+  /// - Precondition: `self` is non-empty.
+  public mutating func popLast() -> T { ... }
+```
+
+A client of this method is considered to have a bug unless
+the array has an element.  OK, so what about postconditions?
+
+The postconditions are the effects of the method plus any returned
+result.  If the preconditions are met, but the postconditions are not,
+and the function does not report an error, we'd say the method has a
+bug.  The bug could be in the documentation of course, *which is a
+part of the method*.
```

Comment:
Is this the first place you say the documentation is part of the method? Clarifying this earlier may resolve some of the discussion between you and @camio regarding "contracts tell us where the bug is"

Reply:
Yeah, as I said in that other thread I think it needs to be made earlier, right after we talk about relationships.

Reply:
OK, I moved the other instance earlier but I really think reminding the reader of that fact in a few words here is necessary.  Otherwise the sentence sounds like it's describing a way in which the bug might not be in the method.